### PR TITLE
Add config to support using client certs for JWKs resolver

### PR DESCRIPTION
**Pull request by ethanchowell**
**Please provide a description of this PR:**
In my system, I have an istio Gateway configured for `MUTUAL` tls and a Keycloak instance sitting behind that gateway. When I try to configure a RequestAuthentication to validate tokens against keycloak using the `jwksUri` istiod is unable to complete this verification because client certs are required. A workaround I've been able to do is apply keycloak configs to essentially hardcode the hostname keycloak thinks it has and point the `jwksUri` to the internal k8s service, but that only applies when keycloak is in the same cluster.

I also have additional systems that require PKI auth through a physical token, and those connections need to be renegotiated to TLSv1.2 #47531. So this PR adds opt-in config for istiod to use a client cert when communicating with a `jwksUri` that required mTLS, as well as opt-in to TLS renegotiation for those endpoints.

### CODE REVIEW

**Comment by zirain** [comment]
can you achieve this with envoy JwksFetchMode?

**Comment by ethanchowell** [comment]
Maybe? If I'm understanding correctly, I'm guessing that would enable me to setup [mTLS origination](https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway-tls-origination/#perform-mutual-tls-origination-with-an-egress-gateway) with an egress-gateway so that istiod (or maybe the sidecar) would use that?

File: pilot/pkg/model/jwks_resolver.go [hunk 1]
```diff
@@ -199,27 +235,24 @@ func newJwksResolverWithCABundlePaths(
...
-				TLSClientConfig: &tls.Config{
-					// nolint: gosec // user explicitly opted into insecure
-					InsecureSkipVerify: features.JwksResolverInsecureSkipVerify,
-					RootCAs:            caCertPool,
-					MinVersion:         tls.VersionTLS12,
-				},
-			},
+	tlsConfig.RootCAs = caCertPool
+	if features.JwksResolverUseClientCerts {
+		cert, err := tls.LoadX509KeyPair(certPath, keyPath)
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/jwks_resolver.go:240]: So it does not support cert update

**Review by howardjohn — COMMENTED** [review]
You could also just use the Envoy resolution mode. JWKS requests come from Envoy, which can use standard Istio configuration like Destinationrule.

cc @liwenhao0810 @aryan16

**Comment by ethanchowell** [comment]
> You could also just use the Envoy resolution mode. JWKS requests come from Envoy, which can use standard Istio configuration like Destinationrule.

If what I'm trying to do here can be done with configuration I'm totally fine with that. So I circled back and am trying to make this happen. I added these values to istiod (helm install)

```yaml
pilot:
  env:
    PILOT_JWT_ENABLE_REMOTE_JWKS: "envoy"
```

Deployed an egressgateway with mTLS origination following [this guide](https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway-tls-origination/#perform-mutual-tls-origination-with-an-egress-gateway) and confirmed that I could jump into a pod and curl a mTLS protected endpoint like `curl -vvv -ksfL http://my-host/keycloak/realms/my-realm/.well-known/openid-configuration`. So then I'm left with a RequestAuthentication like

```yaml
apiVersion: security.istio.io/v1
kind: RequestAuthentication
metadata:
  name: httpbin
spec:
  jwtRules:
    - forwardOriginalToken: true
      fromHeaders:
        - name: x-auth-request-access-token
      issuer: https://my-realm/keycloak/realms/my-realm
      jwksUri: http://my-host/keycloak/realms/my-realm/protocol/openid-connect/certs
  selector:
    matchLabels:
      app.kubernetes.io/instance: httpbin
      app.kubernetes.io/name: httpbin
```

Plus an AuthorizationPolicy to just check a group. I can definitely see that the sidecar for httpbin is trying to reach out

```text
[2023-11-24T23:10:02.559Z] "GET /headers HTTP/1.1" 401 - jwt_authn_access_denied{Jwks_remote_fetch_is_failed} - "-" 0 27 1236 - "10.42.4.0" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36" "72f24a46-605e-495c-9ba6-92cb4357811c" "my-host" "-" inbound|8080|| - 10.42.5.16:8080 10.42.4.0:0 outbound_.8080_._.httpbin.ehowell.svc.cluster.local -
```

But I can never get past the `Jwks remote fetch is failed` error and I don't see anything in the egressgateway logs indicating that it's reaching out. I also tried copying the Secret with the certs, and the Destination rule from that guide (originate-mtls-...) to the same namespace as the sidecar with no luck, same for setting `exportTo[0]='*'` on the original destination rule. Hopefully there's something obvious I'm missing? Admittedly, I haven't yet built and deployed the changes from this PR but that's my next steps assuming I can't get the config worked out.

**Pull request closed without merging**